### ISSUE TITLE
PTX math instructions are wrapped in a function

### ISSUE THREAD

**Issue by xianghao-wang** [chapel-lang/chapel#22112]
### Summary of Problem
When calling math functions from `AutoMath` module, in the generated ptx codes, `sqrt`, `sin`, `cos` and etc. will be wrapped in a PTX function e.g. `__nv_sqrt`. And the call of `__nv_sqrt` could impact on performance. In constrast, CUDA C generated PTX codes directly uses built-in math instruction.
@e-kayrakli 

**Chapel version:** 1.30.0

### Steps to Reproduce

**Source Code of Chapel demo:**

`VectorSquare.chpl`

```chapel
use Random;
use AutoMath;

param N = 1024;
param B = 4;

proc main() {
  on here.gpus[0] {
    var A: [0..<N] real;
    fillRandom(A);

    foreach i in 0..<N / B {
      for param j in 0..<B {
        A[i*B+j] = sqrt(A[i*B+j]);
      }
    }
  }
  
}
```

**Source Code of CUDA C demo:**

`vector_square.cu`

```c
#include <stdio.h>
#include <stdlib.h>

#define N 1024
#define B 4
double *A;
double *A_host;

__global__ void square(double *A) {
  int base = threadIdx.x * B;
  for (int idx = base; idx < base + B; idx ++)
    A[idx] = sqrt(A[idx]);
}

int main(void) {
  srand(0);  
  
  A_host = (double *) malloc(N * sizeof(double));
  
  cudaSetDevice(0);
  cudaMalloc(&A, N);
  cudaDeviceSynchronize();

  cudaMemcpy(A, A_host, N*sizeof(double), cudaMemcpyHostToDevice);  
  square<<<1, N/B>>>(A);
  cudaMemcpy(A_host, A, N*sizeof(double), cudaMemcpyDeviceToHost);  

  return 0;
}
```

**Compile command:**
`chpl --fast --no-ieee-float --savec tmp VectorSquare.chpl`

`nvcc -O3 -use_fast_math --ptx ./vector_square.cu `

**Generated PTX code section of Chapel one:**

```assembly
	// .globl	chpl_gpu_kernel17
.visible .entry chpl_gpu_kernel17(
	.param .u64 chpl_gpu_kernel17_param_0,
	.param .u64 chpl_gpu_kernel17_param_1,
	.param .u64 chpl_gpu_kernel17_param_2,
	.param .u64 chpl_gpu_kernel17_param_3
)
{
	.reg .pred 	%p<4>;
	.reg .b32 	%r<7>;
	.reg .b64 	%rd<21>;
	.reg .f64 	%fd<4>;

	ld.param.u64 	%rd11, [chpl_gpu_kernel17_param_0];
	ld.param.u64 	%rd14, [chpl_gpu_kernel17_param_1];
	mov.u32 	%r1, %ctaid.x;
	mov.u32 	%r2, %ntid.x;
	mov.u32 	%r3, %tid.x;
	mad.lo.s32 	%r4, %r1, %r2, %r3;
	cvt.u32.u64 	%r5, %rd11;
	add.s32 	%r6, %r4, %r5;
	cvt.s64.s32 	%rd2, %r6;
	setp.gt.s64 	%p1, %rd2, %rd14;
	@%p1 bra 	LBB24_4;
	ld.param.u64 	%rd12, [chpl_gpu_kernel17_param_3];
	setp.gt.s64 	%p2, %rd11, %rd12;
	@%p2 bra 	LBB24_4;
	ld.param.u64 	%rd13, [chpl_gpu_kernel17_param_2];
	cvta.to.global.u64 	%rd1, %rd13;
	shl.b64 	%rd3, %rd2, 2;
	ld.global.u64 	%rd15, [%rd1+8];
	ld.u64 	%rd4, [%rd15+144];
	sub.s64 	%rd16, %rd12, %rd11;
	add.s64 	%rd20, %rd16, 1;
	add.s64 	%rd17, %rd11, %rd3;
	shl.b64 	%rd18, %rd17, 3;
	add.s64 	%rd19, %rd4, %rd18;
LBB24_3:
	ld.f64 	%fd1, [%rd19];
	{ // callseq 0, 0
	.reg .b32 temp_param_reg;
	.param .b64 param0;
	st.param.f64 	[param0+0], %fd1;
	.param .b64 retval0;
	call.uni (retval0), 
	__nv_sqrt, 
	(
	param0
	);
	ld.param.f64 	%fd2, [retval0+0];
	} // callseq 0
	st.f64 	[%rd19], %fd2;
	add.s64 	%rd20, %rd20, -1;
	add.s64 	%rd19, %rd19, 8;
	setp.ne.s64 	%p3, %rd20, 0;
	@%p3 bra 	LBB24_3;
LBB24_4:
	ret;

}
	// .globl	__nv_sqrt
.visible .func  (.param .b64 func_retval0) __nv_sqrt(
	.param .b64 __nv_sqrt_param_0
)
{
	.reg .f64 	%fd<3>;

	ld.param.f64 	%fd1, [__nv_sqrt_param_0];
	sqrt.rn.f64 	%fd2, %fd1;
	st.param.f64 	[func_retval0+0], %fd2;
	ret;

}
```

`__nv_sqrt` is a function wrapping `sqrt.rn.f64` instruction.

**Generated PTX code section of CUDA C one:**

```assembly
//
// Generated by NVIDIA NVVM Compiler
//
// Compiler Build ID: CL-31833905
// Cuda compilation tools, release 11.8, V11.8.89
// Based on NVVM 7.0.1
//

.version 7.8
.target sm_52
.address_size 64

	// .globl	_Z6squarePd

.visible .entry _Z6squarePd(
	.param .u64 _Z6squarePd_param_0
)
{
	.reg .b32 	%r<2>;
	.reg .f64 	%fd<9>;
	.reg .b64 	%rd<5>;


	ld.param.u64 	%rd1, [_Z6squarePd_param_0];
	cvta.to.global.u64 	%rd2, %rd1;
	mov.u32 	%r1, %tid.x;
	mul.wide.s32 	%rd3, %r1, 8;
	add.s64 	%rd4, %rd2, %rd3;
	ld.global.f64 	%fd1, [%rd4];
	sqrt.rn.f64 	%fd2, %fd1;
	st.global.f64 	[%rd4], %fd2;
	ld.global.f64 	%fd3, [%rd4+8];
	sqrt.rn.f64 	%fd4, %fd3;
	st.global.f64 	[%rd4+8], %fd4;
	ld.global.f64 	%fd5, [%rd4+16];
	sqrt.rn.f64 	%fd6, %fd5;
	st.global.f64 	[%rd4+16], %fd6;
	ld.global.f64 	%fd7, [%rd4+24];
	sqrt.rn.f64 	%fd8, %fd7;
	st.global.f64 	[%rd4+24], %fd8;
	ret;
}

```
`sqrt.rn.f64` is directly used.

**Comment by e-kayrakli** [comment]
Thanks for the issue and the details @xianghao-wang! This is definitely useful.

A related performance issue (gap between 32- vs 64-bit sqrt) was recorded in an internal note in https://github.com/Cray/chapel-private/issues/4631.

**Comment by bradcray** [comment]
Just noting that this was resolved in #22841.  Thanks @xianghao-wang for reporting the issue, and please let us know if you find others, or find that this one is not resolved once you're using a version of Chapel that includes the fix.
